src/sage/graphs/graph.py

```diff
@@ -5059,11 +5059,377 @@ def is_bicritical(self, matching=None, algorithm='Edmonds', coNP_certificate=Fal
+
+        - A boolean indicating whether the graph is matching covered or not.
+
+        - If ``coNP_certificate`` is set to ``True``, an edge is returned in
+          case the graph is not matching covered otherwise ``None`` is
+          returned.
+
+        EXAMPLES:
+
+        Petersen graph is matching covered::
+
+            sage: G = graphs.PetersenGraph()
+            sage: G.is_matching_covered()
+            True
+
+        A graph is matching covered if and only if the underlying simple graph
+        is matching covered::
+
+            sage: G = graphs.PetersenGraph()
+            sage: G.allow_multiple_edges(True)
+            sage: G.add_edge(0, 5)
+            sage: G.is_matching_covered()
+            True
+
+        A corollary to Tutte's fundamental result [Tut1947]_, as a
+        strengthening of Petersen's Theorem, states that every 2-connected
+        cubic graph is matching covered::
+
+            sage: G = Graph()
+            sage: G.add_edges([
+            ....:    (0, 1), (0, 2), (0, 3),
+            ....:    (1, 2), (1, 4), (2, 4),
+            ....:    (3, 5), (3, 6), (4, 7),
+            ....:    (5, 6), (5, 7), (6, 7)
+            ....: ])
+            sage: G.vertex_connectivity()
+            2
+            sage: degree_sequence = G.degree_sequence()
+            sage: min(degree_sequence) == max(degree_sequence) == 3
+            True
+            sage: G.is_matching_covered()
+            True
+
+        A connected bipartite graph `G[A, B]`, with `|A| = |B| \geq 2`, is
+        matching covered if and only if `|N(X)| \geq |X| + 1`, for all
+        `X \subset A` such that `1 \leq |X| \leq |A| - 1`. For instance,
+        the Hexahedral graph is matching covered, but not the path graphs on
+        even number of vertices, even though they have a perfect matching::
+
+            sage: G = graphs.HexahedralGraph()
+            sage: G.is_bipartite()
+            True
+            sage: G.is_matching_covered()
+            True
+            sage: P = graphs.PathGraph(10)
+            sage: P.is_bipartite()
+            True
+            sage: M = Graph(P.matching())
+            sage: set(P.vertices()) == set(M.vertices())
+            True
+            sage: P.is_matching_covered()
+            False
+
+        A connected bipartite graph `G[A, B]` of order six or more is matching
+        covered if and only if `G - a - b` has a perfect matching for some
+        vertex `a` in `A` and some vertex `b` in `B`::
+
+            sage: G = graphs.CircularLadderGraph(8)
+            sage: G.is_bipartite()
+            True
+            sage: G.is_matching_covered()
+            True
+            sage: A, B = G.bipartite_sets()
+            sage: import random
+            sage: a = random.choice(list(A))                                            # needs random
+            sage: b = random.choice(list(B))                                            # needs random
+            sage: G.delete_vertices([a, b])
```

Comment:
something goes wrong here. variable a should be defined.

Reply:
variables `a` and `b` are created with annotations `# needs random`. Consequently, these variables can be used only for lines with the same annotation.
The solution is to use a block scope, that is add at the beginning of this block of example the line:
```
sage: # needs random
```